pinc/SiteConfig.inc

```diff
@@ -0,0 +1,144 @@
+    public static ?string $forums_json_posts = null;
+
+    // topic forums
+    public static ?int $beginners_site_forum_idx;
+    public static ?int $waiting_projects_forum_idx;
+    public static ?int $projects_forum_idx;
+    public static ?int $pp_projects_forum_idx;
+    public static ?int $posted_projects_forum_idx;
+    public static ?int $deleted_projects_forum_idx;
+    public static ?int $completed_projects_forum_idx;
+    public static ?int $post_processing_forum_idx;
+    public static ?int $teams_forum_idx;
+
+    // proofreading controls
+    public static string $preceding_proofer_restriction = 'not_immediately_preceding';
+    public static bool $public_page_details = false;
+
+    // uploads
+    public static string $uploads_dir;
+    public static string $uploads_subdir_trash = "TRASH";
+    public static string $uploads_subdir_commons = "Commons";
+    public static string $uploads_subdir_users = "Users";
+    public static string $antivirus_executable;
+
+    // WordCheck & locales
+    public static string $aspell_executable = '/usr/bin/aspell';
+    public static string $aspell_prefix = "/usr";
+    public static string $system_locales_dir = '/usr/share/locale';
+
+    // API
+    public static bool $api_enabled = true;
+    public static bool $api_rate_limit = false;
+    public static int $api_rate_limit_requests_per_window = 3600;
+    public static int $api_rate_limit_seconds_in_window = 3600;
+    public static array $api_storage_keys = [];
+
+    // emails
+    public static array $phpmailer_smtp_config = [];
+    public static ?string $no_reply_email_addr;  // only used by noncvs code
+    public static ?string $general_help_email_addr;
+    public static ?string $site_manager_email_addr;
+    public static ?string $auto_email_addr;
+    public static ?string $db_requests_email_addr;
+    public static ?string $promotion_requests_email_addr;
+    public static ?string $ppv_reporting_email_addr;
+    public static ?string $image_sources_manager_addr;
+    public static ?string $translation_coordinator_email_addr;
+
+    // Misc configuration
+    public static bool $use_secure_cookies = false;
+    public static string $php_cli_executable = '/usr/bin/php';
+    public static ?string $site_registration_protection_code = null;
+    public static bool $auto_post_to_project_topic = false;
+    public static bool $ordinary_users_can_see_queue_settings = true;
+    public static string $external_catalog_locator = 'lx2.loc.gov:210/LCDB';
+    public static array $default_project_char_suites = ["basic-latin"];
+    public static bool $testing = false;
+
+    public static function load(string $filename): void
+    {
+        if (SiteConfig::$_is_loaded) {
+            return;
+        }
+
+        try {
+            require $filename;
+        } catch (Error $error) {
+            throw new RuntimeException("Configuration file $filename not found");
+        }
+
+        $ro = new ReflectionObject(new SiteConfig());
```

Comment:
Most languages warn against using reflection as it tends to be unoptimized. Not sure where PHP stands on this and how much we care about it in practice... Thought I will bring it so we're aware of this potential issue.

The alternatives have other downsides:
1. an explicit list (probably with a test to make sure it stays up-to-date)
2. some autogenerator (which is probably going down the Symphony's `Config` route)

Reply:
Some rough performance benchmarking says we don't need to worry about this use of reflection:
```
Reflection: 100000 iterations completed in 8650342746 nanoseconds, or 86503.42746 ns per iteration
Reflection: 100000 iterations completed in 8850424277 nanoseconds, or 88504.24277 ns per iteration

No reflection: 100000 iterations completed in 8419315452 nanoseconds, or 84193.15452 ns per iteration
No reflection: 100000 iterations completed in 8783086792 nanoseconds, or 87830.86792 ns per iteration
```